lib/mini_defender/rules/url.rb

```diff
@@ -1,17 +1,79 @@
 # frozen_string_literal: true
 
 require 'uri'
+require 'resolv'
+require_relative '../detectors/public_domain_detector'
 
 class MiniDefender::Rules::Url < MiniDefender::Rule
+  ALLOWED_MODIFIERS = %w[https_only public no_ip]
+
+  def initialize(modifiers = [])
+    @modifiers = Array(modifiers).map(&:to_s)
+    validate_modifiers! unless @modifiers.empty?
+  end
+
   def self.signature
     'url'
   end
 
+  def self.make(args)
+    new(args)
+  end
+
   def passes?(attribute, value, validator)
-    value.is_a?(String) && URI.regexp(%w[http https]).match?(value)
+    # TODO: warning: URI.regexp is obsolete; use URI::DEFAULT_PARSER.make_regexp instead
+    return false unless value.is_a?(String) && URI.regexp(%w[http https]).match?(value)
+
+    begin
+      uri = URI.parse(value)
+
+      return true if @modifiers.empty?
+
+      return false if @modifiers.include?('https_only') && uri.scheme != 'https'
+      return false if @modifiers.include?('public') &&
+                      !MiniDefender::Detectors::PublicDomainDetector.public_domain?(uri.host)
+      return false if @modifiers.include?('no_ip') && ip_address?(uri.host)
+
+      true
+    rescue URI::InvalidURIError
+      false
+    end
   end
 
-  def message(attribute, value, validator)
-    'The field must contain a valid URL.'
+  def message(_attribute, value, _validator)
+    return 'The field must contain a valid URL.' unless value.is_a?(String)
+
+    begin
+      uri = URI.parse(value)
+      return 'The field must contain a valid URL.' if @modifiers.empty?
```

Comment:
Add a state variable in `passes?` that will hold the error reason to avoid doing validation logic again here

Reply:
Noted 🌟